### PR TITLE
Switch over framework packaging to use Crossgen2

### PR DESCRIPTION
**Pull request by trylek**
Based on Jeremy's advice I have modified the pack build projects
to use the Crossgen2 compiler. As we're still using the 5.0 version
of the SDK, this change synthesizes the name of the JIT library -
upgrading SDK to version 6 will facilitate further cleanup in
these scripts.

Thanks

Tomas

P.S. I'm not sure whether we should add crossgen2 to 'RuntimeFiles' - I originally did that but the packs build complains the "public key not found" in the managed crossgen2 assemblies. I would naively assume that we're publishing Crossgen2 as a separate nuget package so we shouldn't need to include in runtime too.

/cc @dotnet/crossgen-contrib

### CODE REVIEW

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj [hunk 1]
```diff
@@ -18,6 +18,7 @@
     <PublishReadyToRun>true</PublishReadyToRun>
     <!-- Disable crossgen on FreeBSD, NetBSD, illumos and Solaris for now. This can be revisited when we have full support. -->
     <PublishReadyToRun Condition="'$(TargetOS)'=='FreeBSD' Or '$(TargetOS)'=='NetBSD' Or '$(TargetOS)'=='illumos' Or '$(TargetOS)'=='Solaris'">false</PublishReadyToRun>
+    <PublishReadyToRunUseCrossgen2>true</PublishReadyToRunUseCrossgen2>
```
Comment (mangod9) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:21]: so does this use the 5.0 crossgen2 or it will still use the live bits for building?

Comment (jkoritzinsky) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:21]: This uses the live bits for building. The ResolveReadyToRunCompilers target below is how the SDK determines which crossgen2 to use.

Comment (mangod9) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:21]: ok thanks.

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj [hunk 2]
```diff
@@ -74,11 +75,35 @@
   <Import Project="Sdk.targets" Sdk="Microsoft.DotNet.SharedFramework.Sdk" />
 
   <Target Name="ResolveReadyToRunCompilers" DependsOnTargets="ResolveRuntimeFilesFromLocalBuild">
+    <PropertyGroup> <!-- Can be removed once runtime repo switches over to 6.0 drop of the SDK repo -->
+      <JitTargetArchComponent>$(TargetArchitecture)</JitTargetArchComponent>
+      <JitTargetOSComponent Condition="'$(TargetOS)' == 'windows'">win</JitTargetOSComponent>
+      <JitTargetOSComponent Condition="'$(TargetOS)' == 'linux'">unix</JitTargetOSComponent>
+      <JitTargetOSComponent Condition="'$(TargetOS)' == 'osx'">osx</JitTargetOSComponent>
```
Comment (AntonLapounov) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:82]: I think we never use `osx`, do we? See crossgen2.csproj: https://github.com/dotnet/runtime/blob/69e114c1abf91241a0eeecf1ecceab4711b8aa62/src/coreclr/tools/aot/crossgen2/crossgen2.csproj#L64-L65

Comment (trylek) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:82]: Hmm, you're right for sure, we should produce identical code when targeting Linux and OSX. In light of this I'm wondering why I'm seeing "clrjit_unit_osx_arm64_x64.dll" in my crossgen2 folder - @davidwrighton, can you please explain to me what that stands for?

Comment (AntonLapounov) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:82]: At present we use `unix_osx` for OSX on ARM64. You can ignore that for now. #45467

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj [hunk 3]
```diff
@@ -74,11 +75,35 @@
...
+      <JitBuildArchComponent>$(BuildArchitecture)</JitBuildArchComponent>
+      <JitArchSuffix>$(JitTargetArchComponent)_$(JitBuildArchComponent)</JitArchSuffix>
+      <JitFileName>clrjit_$(JitTargetOSComponent)_$(JitArchSuffix)</JitFileName>
+    </PropertyGroup>
+
+    <PropertyGroup Condition="'$(TargetOS)' == 'osx'">
+      <JitFileNameOSX>clrjit_unix_osx_$(JitArchSuffix)</JitFileNameOSX>
+      <JitFileName Condition="Exists($(JitFileNameOSX))">$(JitFileNameOSX)</JitFileName>
+    </PropertyGroup>
```
Comment (AntonLapounov) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:91]: Based on the previous comment, I think this should be removed.

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj [hunk 4]
```diff
@@ -74,11 +75,35 @@
...
+
+    <PropertyGroup>
+      <JitExtension>.so</JitExtension>
+      <JitExtension Condition="'$([System.Runtime.InteropServices.RuntimeInformation]::IsOSPlatform($([System.Runtime.InteropServices.OSPlatform]::OSX)))' == 'true'">.dylib</JitExtension>
+      <JitExtension Condition="'$(OS)' == 'Windows_NT'">.dll</JitExtension>
```
Comment (AntonLapounov) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:96]: Can we use `[MSBuild]::IsOsPlatform('OSX')` instead? See https://github.com/dotnet/runtimelab/blob/feature/NativeAOT/src/coreclr/tools/aot/ILCompiler/ILCompiler.csproj#L24-L27

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj (hunk 2 above)
Comment (trylek) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj]: For your primary comment, I have addressed that in 2nd commit along with a few additional bugs on Linux as I traditionally debugged this on Windows first ;-).

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj (hunk 4 above)
Comment (trylek) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj]: Removed in 2nd commit.

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj [hunk 5]
```diff
@@ -74,11 +75,33 @@
   <Import Project="Sdk.targets" Sdk="Microsoft.DotNet.SharedFramework.Sdk" />
 
   <Target Name="ResolveReadyToRunCompilers" DependsOnTargets="ResolveRuntimeFilesFromLocalBuild">
+    <PropertyGroup> <!-- Can be removed once runtime repo switches over to 6.0 drop of the SDK repo -->
+      <JitTargetArchComponent>$(TargetArchitecture)</JitTargetArchComponent>
+      <JitTargetOSComponent>unix</JitTargetOSComponent>
+      <JitTargetOSComponent Condition="'$(TargetOS)' == 'windows'">win</JitTargetOSComponent>
+      <JitBuildArchComponent>$(BuildArchitecture)</JitBuildArchComponent>
+      <JitArchSuffix>$(JitTargetArchComponent)_$(JitBuildArchComponent)</JitArchSuffix>
+      <JitFileName>$(LibPrefix)clrjit_$(JitTargetOSComponent)_$(JitArchSuffix)</JitFileName>
```
Comment (AntonLapounov) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:84]: I would avoid introducing single-use properties:
```suggestion
      <JitTargetOSComponent>unix</JitTargetOSComponent>
      <JitTargetOSComponent Condition="'$(TargetOS)' == 'windows'">win</JitTargetOSComponent>
      <JitFileName>$(LibPrefix)clrjit_$(JitTargetOSComponent)_$(TargetArchitecture)_$(BuildArchitecture)</JitFileName>
```

Comment (trylek) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:84]: Sounds good, will fix in a bit, I just don't want to restart the build now as I want to see the effect of the Unix change on the build (whether at least something passes now).

Comment (trylek) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj]: Fixed in 3rd commit.

Comment (trylek) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj]: (Along with a few related cleanups.)

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj (hunk 3 above)
Comment (trylek) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj]: Removed in 4th commit. I would still love to hear from @davidwrighton what is the purpose of the file "clrjit_unix_osx_arm64_x64.dll" in my Windows x64 build.

Comment (AntonLapounov) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj]: It is a temporary JIT library for targeting OSX ARM64, which will be replaced by a runtime switch, see #45501.

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj [hunk 6]
```diff
@@ -75,20 +75,29 @@
   <Import Project="Sdk.targets" Sdk="Microsoft.DotNet.SharedFramework.Sdk" />
 
   <Target Name="ResolveReadyToRunCompilers" DependsOnTargets="ResolveRuntimeFilesFromLocalBuild">
-    <PropertyGroup> <!-- Can be removed once runtime repo switches over to 6.0 drop of the SDK repo -->
+    <!-- The following property groups can be simplified once runtime repo switches over to SDK 6.0 drop -->
+    <PropertyGroup Condition="'$(BuildArchitecture)' != '$(TargetArchitecture)'">
+      <CrossDir Condition="'$(TargetArchitecture)' == 'arm' or '$(TargetArchitecture)' == 'arm64'">x64</CrossDir>
+      <CrossDir Condition="'$(TargetArchitecture)' == 'x86'">$(BuildArchitecture)</CrossDir>
```
Comment (AntonLapounov) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:81]: What is different for x86 target architecture?

Comment (trylek) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:81]: https://github.com/dotnet/runtime/issues/47388

Comment (AntonLapounov) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:81]: And what value do you expect `$(BuildArchitecture)` to be equal to for x86?

Comment (trylek) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:81]: I believe that BuildArchitecture is the "natural" architecture of the build machine, on my box it returns "x64" and that works fine.

Comment (trylek) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:81]: It's basically an equivalent of what we have in the SPC crossgenning script,

https://github.com/dotnet/runtime/blob/9f8aab73d93156933ae65a476204bf62c02f6537/src/coreclr/crossgen-corelib.proj#L34

Comment (AntonLapounov) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:81]: Right, it will be x64. Then back to my question: why do we treat x86 differently here?  We could merge those two lines and treat all architectures (arm, arm64, x86) uniformly.

Comment (trylek) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:81]: Hmm, I was under the impression that it would break the build on a machine that's actually x86 but I guess that runtime repo may have trouble building there anyway.

Comment (AntonLapounov) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:81]: These lines are under `Condition="'$(BuildArchitecture)' != '$(TargetArchitecture)'"`.  So if `TargetArchitecture` is x86, then `BuildArchitecture` cannot be x86.

Comment (trylek) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:81]: OK, that's true, I'll change it as you suggest then, thanks for pointing that out.

Comment (trylek) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj]: Fixed in 6th commit.

Comment (AntonLapounov) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj]: Thank you!

**Review by AntonLapounov — APPROVED** [review]

**Comment by trylek** [comment]
Hmm, so on Linux we apparently cannot run the "exe" version of Crossgen2 because it fails to locate the framework. @vitek-karas - is there any magic environment variable we could use to present the Crossgen2 app with the dotnet host installed by Arcade in the repo root? Otherwise we'll need to switch over to using the dotnet launcher and that will require changes even in the 6.0 SDK as its current main branch doesn't have that support.

**Comment by vitek-karas** [comment]
Set `DOTNET_ROOT` to point to the root of the "install". In case of repo local it typically looks something like `/path/to/repo/.dotnet`.

**Comment by jkotas** [comment]
I think you want to use `DOTNET_ROOT` on Windows too. It works on Windows because of it picks up a random version of dotnet that comes with Visual Studio. We try to always use our own local version of dotnet.

**Comment by trylek** [comment]
I believe I have addressed @jkotas' and @jkoritzinsky's PR feedback and the installer tests are now passing. I'm however seeing a few library test failures, most notably <code>System.Linq.Expressions.Tests.ConvertTests.ConvertFloatToUIntTest</code> and <code>System.Linq.Expressions.Tests.ConvertTests.ConvertFloatToNullableUIntTest</code>, these repro on several platforms and indicate a potentially systemic failure. I plan to look deeper into these tests next unless someone has another explanation for the failures.

**Comment by trylek** [comment]
In fact I have a hard time to imagine how my "packs.product"-only changes could affect library tests but it's been a while since I was buried full time in runtime repo consolidation and things may have changed since then.

File: eng/build.sh [hunk 7]
```diff
@@ -14,6 +14,8 @@ while [[ -h "$source" ]]; do
 done
 scriptroot="$( cd -P "$( dirname "$source" )" && pwd )"
 
+export DOTNET_ROOT=$scriptroot/../.dotnet
```
[the remark this answers is not in the document]
Comment (trylek) [reply eng/build.sh]: Thanks for pointing that out, hopefully addressed in the 8th commit. Fixing this cleanly also unblocked OSX runs that were failing with the above band-aid fix.

**Comment by jkoritzinsky** [comment]
>I'm however seeing a few library test failures, most notably `System.Linq.Expressions.Tests.ConvertTests.ConvertFloatToUIntTest` and `System.Linq.Expressions.Tests.ConvertTests.ConvertFloatToNullableUIntTest`, these repro on several platforms and indicate a potentially systemic failure. I plan to look deeper into these tests next unless someone has another explanation for the failures.

Those failures are known failures.

File: eng/common/build.sh [hunk 8]
```diff
@@ -177,6 +177,8 @@ if [[ "$ci" == true ]]; then
 fi
 
 . "$scriptroot/tools.sh"
+InitializeDotNetCli true
+export DOTNET_ROOT=$_InitializeDotNetCli
```
Comment (ViktorHofer) [inline eng/common/build.sh:181]: undo these changes as they are mirrored by arcade and will be overwritten

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj [hunk 9]
```diff
@@ -74,6 +75,21 @@
   <Import Project="Sdk.targets" Sdk="Microsoft.DotNet.SharedFramework.Sdk" />
 
   <Target Name="ResolveReadyToRunCompilers" DependsOnTargets="ResolveRuntimeFilesFromLocalBuild">
+    <!-- The following property group can be simplified once runtime repo switches over to SDK 6.0 drop -->
+    <PropertyGroup>
+      <CrossDir></CrossDir>
```
Comment (ViktorHofer) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:80]: nit:
```suggestion
      <CrossDir />
```

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj [hunk 10]
```diff
@@ -125,6 +126,21 @@
   <Import Project="Sdk.targets" Sdk="Microsoft.DotNet.SharedFramework.Sdk" />
 
   <Target Name="ResolveReadyToRunCompilers" DependsOnTargets="ResolveRuntimeFilesFromLocalBuild">
+    <!-- The following property group can be simplified once runtime repo switches over to SDK 6.0 drop -->
```
Comment (ViktorHofer) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj:129]: Looks like these properties are a duplicate of the ones in the touched file above. Would it make sense to extract them into a props file and import them in both projects to avoid the duplication?

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj (hunk 9 above)
Comment (trylek) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:80]: Oh, sure, good point, thanks for spotting.

File: eng/build.ps1 [hunk 11]
```diff
@@ -120,6 +120,9 @@ if ($subset -eq 'help') {
   exit 0
 }
 
+# This tells .NET Core to use the bootstrapped runtime
+$env:DOTNET_ROOT=InitializeDotNetCli -install:$true -createSdkLocationFile:$true
```
Comment (ViktorHofer) [inline eng/build.ps1:124]: Why do you need DOTNET_ROOT to be set? We didn't need this before except for Visual Studio and ILLink (?)?

Comment (trylek) [reply eng/build.ps1:124]: On Linux, Crossgen2 executable doesn't launch without DOTNET_ROOT being set (please discussion above in this thread). I guess the clean solution would be to use the dotnet launcher but that's something yet to be implemented in the sdk repo so even after its upgrade to 6.0 it won't come for free.

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj (hunk 10 above)
Comment (trylek) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj:129]: I agree that's a straightforward copy but I suspect it can't be props as it's a target with a state. I think we could make it an Import but frankly speaking 90% of the code will be deleted after Crossgen1 and SDK 5.0 are retired so I'm not sure if it's worth the effort, ultimately the remaining Crossgen2 variant should boil down to less than half dozen lines of msbuild script code.

File: eng/build.ps1 (hunk 11 above)
Comment (ViktorHofer) [reply eng/build.ps1:124]: Can we set this environment variable only in the place where crossgen2 is actually invoked? Is that happening in the SDK? Is there a hook to pass environment variable to that invocation directly to avoid global environment variables being set? An MSBuild Exec Task allows to pass in env vars.

**Comment by trylek** [comment]
OK, so this is my roster of items to follow up for now:

1) SDK repo drop in the runtime repo needs upgrading to 6.0 (will remove the need to synthesize JitPath in the scripts);
2) SDK repo implementation of RunReadyToRunCompiler lacks actionable telemetry, cf. https://github.com/dotnet/sdk/issues/15466
3) Microsoft.NETCore.App.Crossgen2.sfxproj and Microsoft.NETCore.App.Runtime.sfxproj contain a duplicative section related to R2R compilation (Crossgen / Crossgen2 configuration); most of this should go away once SDK 5.0 and Crossgen1 are retired.

Please remind me if I missed something.

File: eng/common/build.sh (hunk 8 above)
Comment (trylek) [reply eng/common/build.sh:181]: Hmm, OK, I guess that means that I should make the change in a different place, otherwise I'll need to make it in Arcade.

File: eng/build.ps1 (hunk 11 above)
Comment (ViktorHofer) [reply eng/build.ps1:124]: We don't want to set this environment variable in the root build script as it means that we depend on it throughout the whole build which we actually don't as only crossgen2 needs it. Setting it in the place where crossgen2 is invoked guarantees, that the sfxprojs are buildable when that env var isn't set (i.e. when invoked directly with `dotnet build`, instead of from the root build.cmd/sh).

Comment (trylek) [reply eng/build.ps1:124]: Well, I guess it's mostly a sequencing problem. My initial goal was to just make this work somehow; with help from many people I've managed to achieve that in the current change and we can now discuss the logistics of getting that in. I believe that the problem regarding <code>DOTNET_ROOT</code> is basically a deficiency of the <code>RunReadyToRunCompiler</code> task that should accept a parameter specifying the path to .NET CLI and use the "dotnet" launcher to launch Crossgen2. This logic is absent both in SDK 5.0 and in current main branch of the repo i.o.w. it's not available to us now and won't become immediately available even after we upgrade the SDK repo drop. I can easily sit on this change for a week or two and make the prerequisite changes first in other repos but we need to coordinate the integrations to make sense.

Comment (trylek) [reply eng/build.ps1:124]: In case this is a too oblique answer to your question, the <code>ToolTask</code> base class of <code>RunReadyToRunCompiler</code>, please see

https://github.com/dotnet/sdk/blob/810129ba5b35dbf77f8d6d84171eb7363693551d/src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs#L14

currently doesn't support injecting environment variable overrides into the child process. It is definitely doable, we're doing such a thing in R2RTest and in other places, it's just not there yet. This gets us back to the sequencing thing, it seems to me we're converging towards the conclusion that the best way forward is to fix this cleanly in the SDK repo and simplify this change to just consume it, in such case we'd no longer require global environment overrides like <code>DOTNET_ROOT</code>.

Comment (trylek) [reply eng/build.ps1:124]: Actually the formulation I used is somewhat imprecise - I don't claim ToolTask doesn't support setting environment variables, I have no idea if it does and maybe / likely it does, it's just not used / implemented in the RunReadyToRunCompiler task class derived from ToolTask and so implementing it would require a code change. And my current thinking is that, if we're about to do a code change to the RunReadyToRunCompiler task, it would be easiest to cleanly switch it over to running Crossgen2 using the dotnet launcher and avoid tampering with environment variables.

Comment (ViktorHofer) [reply eng/build.ps1:124]: in response to your ToolTask statement, ToolTask already provides a way to specify env vars: https://docs.microsoft.com/en-us/dotnet/api/microsoft.build.utilities.tooltask.environmentvariables?view=msbuild-16-netcore#Microsoft_Build_Utilities_ToolTask_EnvironmentVariables. It's probably just a matter of adding an environment variables property to the RunReadyToRunCompiler task invocation.

Regarding the logistics, I agree that this should be cleanly fixed by the SDK but that doesn't mean that the required environment variable can't be set just in time before the RunReadyToRunCompiler task is invoked.

I was thinking about adding a target that hooks onto `_CreateR2RImages` as a BeforeTargets and sets the environment variable.

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj (hunk 10 above)
Comment (ViktorHofer) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj:129]: If this is just temporary then I'm fine with the duplication

**Comment by trylek** [comment]
It looks like I managed to pass installer builds using the modification per Viktor's suggestion. @jkoritzinsky / @ViktorHofer, can you please take another look when you have a chance?

Thanks

Tomas

File: src/installer/pkg/sfx/Microsoft.NETCore.App/ReadyToRun.targets [hunk 12]
```diff
@@ -16,5 +31,34 @@
...
+  <Target Name="PatchDotnetRootBeforeRunningCrossgen2" BeforeTargets="_CreateR2RImages">
+    <PropertyGroup>
+      <OriginalDotnetRootValue>$(DOTNET_ROOT)</OriginalDotnetRootValue>
+      <DOTNET_ROOT>$(RepoRoot)</DOTNET_ROOT>
+    </PropertyGroup>
+  </Target>
+
+  <Target Name="RestoreDotnetRootAfterRunningCrossgen2" AfterTargets="_CreateR2RImages">
+    <PropertyGroup>
+      <DOTNET_ROOT>$(OriginalDotnetRootValue)</DOTNET_ROOT>
```
Comment (hoyosjs) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/ReadyToRun.targets:57]: The only problem with this is it will eventually change this property, some time after _CreateR2RImages ran. The only 2 ways I know for somewhat strict ordering are using a dependent target (the you don't own case, but there's an extensibility model, and it's cleaner), or the CallTarget route (which has few good uses, and unless we own _CreateR2RImages, it won't be a good idea..

Comment (trylek) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/ReadyToRun.targets:57]: The exact problem here is that, while we "own" _CreateR2RImages, we do so in a different repo (sdk) and the entire trick here is about not having to wait for its propagation to the runtime repo as that will take about two months or more. If I could merge in my SDK PR today and propagate it to runtime tomorrow, this would be a no-concern.

File: eng/common/build.sh (hunk 8 above)
Comment (trylek) [reply eng/common/build.sh]: Removed.

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj (hunk 10 above)
Comment (trylek) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj]: Removed duplication with https://github.com/dotnet/runtime/pull/48201.

File: eng/build.ps1 (hunk 11 above)
Comment (trylek) [reply eng/build.ps1]: Good point, fixed.

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj (hunk 9 above)
Comment (trylek) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj]: Fixed in 2nd commit (in ReadyToRun.targets where I moved the logic that was originally duplicated between the two sfxproj scripts).

**Review by ViktorHofer — APPROVED** [review]
LGTM if this works as expected. We will upgrade the SDK in less than two weeks anyway so I'm not too concerned about the hacks. And after that we will update the SDK probably on a monthly cadence.

**Comment by AntonLapounov** [comment]
Apparently Crossgen2 is unable to emit PDBs on ARM64.  We would need to grab `Microsoft.DiaSymReader.Native.arm64.dll` for that.
```
C:\Runtime\dotnet.cmd C:\Runtime\artifacts\bin\coreclr\windows.arm64.Release\crossgen2\crossgen2.dll -o:C:\Runtime\artifacts\bin\coreclr\windows.arm64.Release\System.Private.CoreLib.dll -r:C:\Runtime\artifacts\bin\coreclr\windows.arm64.Release\IL\*.dll --targetarch:arm64 -O C:\Runtime\artifacts\bin\coreclr\windows.arm64.Release\IL\System.Private.CoreLib.dll --pdb --pdb-path:C:\Runtime\artifacts\bin\coreclr\windows.arm64.Release\PDB
  C:\Runtime\.dotnet
  Emitting R2R PE file: C:\Runtime\artifacts\bin\coreclr\windows.arm64.Release\System.Private.CoreLib.dll
  Emitting PDB file: C:\Runtime\artifacts\bin\coreclr\windows.arm64.Release\PDB\System.Private.CoreLib.ni.pdb
EXEC : error : An attempt was made to load a program with an incorrect format. (0x8007000B) [C:\Runtime\src\coreclr\crossgen-corelib.proj]
  System.BadImageFormatException: An attempt was made to load a program with an incorrect format. (0x8007000B)
     at ILCompiler.Diagnostics.PdbWriter.CreateNGenPdbWriter64(String ngenImagePath, String pdbPath, Object& ngenPdbWriter)
```
